src/segment-loader.js

```diff
@@ -287,28 +315,74 @@
 
       // stop processing if the request was aborted
       if (!request.response) {
+        this.abort_();
         return;
       }
 
-      // calculate the download bandwidth
-      this.roundTrip = request.roundTripTime;
-      this.bandwidth = request.bandwidth;
-      this.bytesReceived += request.bytesReceived || 0;
+      if (request === this.xhr_.segmentXhr) {
+        segmentXhrRequest = this.xhr_.segmentXhr;
+        // the segment request is no longer outstanding
+        this.xhr_.segmentXhr = null;
 
-      segment = segmentInfo.playlist.segments[segmentInfo.mediaIndex];
-      if (segment.key) {
-        segmentInfo.encryptedBytes = new Uint8Array(request.response);
-      } else {
-        segmentInfo.bytes = new Uint8Array(request.response);
+        // calculate the download bandwidth based on segment request
+        this.roundTrip = request.roundTripTime;
+        this.bandwidth = request.bandwidth;
+        this.bytesReceived += request.bytesReceived || 0;
+
+        if (segment.key) {
+          segmentInfo.encryptedBytes = new Uint8Array(request.response);
+        } else {
+          segmentInfo.bytes = new Uint8Array(request.response);
+        }
       }
 
-      this.processResponse_();
+      if (request === this.xhr_.keyXhr) {
+        keyXhrRequest = this.xhr_.segmentXhr;
+        // the key request is no longer outstanding
+        this.xhr_.keyXhr = null;
+
+        view = new DataView(request.response);
+        segment.key.bytes = new Uint32Array([
+          view.getUint32(0),
+          view.getUint32(4),
+          view.getUint32(8),
+          view.getUint32(12)
+        ]);
+      }
+
+      if (!this.xhr_.segmentXhr && !this.xhr_.keyXhr) {
+        this.xhr_ = null;
+        this.processResponse_();
+      }
     },
 
     processResponse_: function() {
+      var segmentInfo, playlist, segment, mediaIndex, segIv;
```

Comment:
Not in love with the name `segIv` but I know this was pulled over from the old code.